csharp/src/Ice/VectoredBufferExtensions.cs

```diff
@@ -0,0 +1,135 @@
+                    i++;
+                    break;
+                }
+                else
+                {
+                    srcOffset -= segment.Count;
+                }
+            }
+
+            for (; i < src.Count && sz < count; i++)
+            {
+                ArraySegment<byte> segment = src[i];
+                if (segment.Count > count - sz)
+                {
+                    dst.Add(segment.Slice(0, count - sz));
+                    break;
+                }
+                else
+                {
+                    dst.Add(segment);
+                    sz += segment.Count;
+                }
+            }
+        }
+
+        /// <summary>Returns and array segment with the requested bytes, starting at the given
+        /// byte offset, if the required bytes are available in a single segment this method returns
+        /// an slice of the segment without copying the data, otherwise the data is copied into
+        /// a byte array and the segment is created from the copied data.</summary>
+        /// <param name="src">The source segment list.</param>
+        /// <param name="offset">The zero-based byte offset into the source segment list.</param>
+        /// <param name="count">The size in bytes of the returned segment.</param>
+        /// <returns>A segment with the requested size.</returns>
+        public static ArraySegment<byte> GetSegment(this IList<ArraySegment<byte>> src, int offset, int count)
+        {
+            Debug.Assert(src.GetByteCount() >= offset + count,
+                $"requested {count} bytes starting at offset {offset} but there is only " +
+                $"{src.GetByteCount() - offset} bytes remaining.");
+
+            // Skip offset bytes into the source segment list
+            int remaining;
+            int i = 0;
+            byte[]? data = null;
+            for (; i < src.Count; i++)
+            {
+                ArraySegment<byte> segment = src[i];
+                if (segment.Count > offset)
+                {
+                    if (segment.Count - offset >= count)
+                    {
+                        // If the requested data is available from a single segment return an Slice
+                        // of the segment.
+                        return segment.Slice(offset, count);
+                    }
+                    else
+                    {
+                        // The requested data spams serveral segments, allocate an array of the requested
+                        // size and copy the data into it. We first copy the remainig of the current segment
+                        // here and the rest of the data is copied in the loop bellow.
+                        data = new byte[count];
+                        remaining = segment.Count - offset;
+                        Buffer.BlockCopy(segment.Array, offset, data, segment.Offset, remaining);
+                        offset = remaining;
```

Comment:
Should this be += remaining?

Reply:
No, here we are setting offset to the dstOffset for the next BlockCopy in the loop below

Reply:
I am not following. offset represents the absolute offset is the src buffer. remaining is the number of bytes we just wrote into the destination buffer. Sounds like we should increase offset by remaining.

Reply:
I think it is confusing because it is first used as the offset into the src, and then as the ofset into dst array, I can add a second dstOffset to make this clear.

Reply:
I see now. Yes, another variable would be better.